### ISSUE TITLE
[ogbv plugin] Accessibility

### ISSUE THREAD

**Issue by duggalsu** [tattle-made/Uli#77]
# General

Currently tested on
- [x] Ubuntu 20.04, Orca screen reader, Gnome desktop
- [x] OGBV plugin - v.0.0.9

Languages tested
- [x] English
- [ ] Hindi
- [ ] Tamil

Currently NOT testing
- [ ] Braille - Lack of hardware resource

Development and Testing
https://reactjs.org/docs/accessibility.html#development-and-testing-tools

References
- https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML
- https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#screenreaders
- https://webaim.org/techniques/screenreader/
- https://wiki.gnome.org/Projects/Orca
- https://developer.chrome.com/docs/extensions/mv3/a11y/
- https://www.chromium.org/developers/design-documents/accessibility/#TOC-WAI-ARIA-Support
- https://v2.grommet.io/
- https://v1.grommet.io/docs/accessibility
- https://github.com/grommet/grommet/wiki/2.0-Migration-Guide
- https://reactjs.org/docs/accessibility.html

---

# TODO

## Chrome Extensions

- [x] Accessible UI controls
https://developer.chrome.com/docs/extensions/mv3/a11y/#controls
https://developer.chrome.com/docs/extensions/mv3/a11y/#htmlcontrols

*Only standard controls used*

- [ ] Keyboard navigation
https://developer.chrome.com/docs/extensions/mv3/a11y/#navigation
https://support.google.com/chrome/answer/157179
```
Set focus on the rightmost item in the Chrome toolbar	F10 
```

https://www.webnots.com/twitter-keyboard-shortcuts/
```
When you on the Twitter webpage, press the ? key (shift + ?) on your keyboard to get the full list of keyboard shortcuts.
```

TODO
- 'Ask friends for help' - Popup navigation issue 

[Edit 1] Refer - https://github.com/grommet/grommet/issues/6045

If no options are selected in the popup and the popup loses focus
- `esc` key does not work
- focus is never returned to popup
- OGBV menu items on twitter page are not selected on keyboard navigation 
https://developer.chrome.com/docs/extensions/mv3/user_interface/#popup

- https://v2.grommet.io/layer
Options
```
modal
Whether there should be an overlay preventing interaction underneath the layer.
```

- https://github.com/grommet/grommet/issues/4721
```
When Layer component is set up as a modal, it still allows keyboard focus to go behind the modal.

Expected Behavior
I would expect that when a modal is open, keyboard navigation should not leave the modal (except Esc which should close the modal). This is supported by the description of the modal attribute in the doc: "Whether there should be an overlay preventing interaction underneath the layer."
I'd expect that continuing to Tab through the modal should loop through the focusable elements inside it.

Actual Behavior
When a modal is open, tabbing allows you to leave the modal and select elements behind it.
```

- [x] Shortcuts
https://developer.chrome.com/docs/extensions/mv3/a11y/#shortcuts

*None implemented*

- [x] Text
https://developer.chrome.com/docs/extensions/mv3/a11y/#text
```
Font choices and text size impact how readable an extension's content is. Users with sight issues may need to increase an extensions text size. If using keyboard shortcuts, make sure they do not interfere with the zoom shortcuts built into Chrome.

As an indicator of flexibility of an extension's UI, apply the 200% test; if the text size or page zoom is increased 200%, is it still usable?

Avoid baking text into images. Users are unable to modify the size and screen readers are unable to interpret images. Instead, opt for styled web font, such as one of the fonts found in the Google Font API. Web fonts can scale to different sizes and can be accessed by people using screen readers.
```

https://www.w3.org/TR/2008/REC-WCAG20-20081211/#visual-audio-contrast-scale
Testing instructions:
- Settings (3 dots) > Font size (+/-)
	- Extension text size remains the same
	- Keyboard navigation works properly
- Increase text size: Settings > Appearance > Page zoom - set to 200%
	- Increases text size in extension too
	- Keyboard navigation works properly
- No images used with text

- [ ] Color
*Currently being modified by designers*

Test all colors with Twitter color schemes
- `Default` background: #FFF
- `Dim` background: #15202b
- `Lights out` background: #000

https://developer.chrome.com/docs/devtools/accessibility/reference/#contrast

TODO
- Verify how to handle this
  - Chrome dev tools dom for twitter webpage buttons - 'Archive' + 'Ask friends for help' - no contrast info available

https://webaim.org/articles/contrast/
https://webdesign.tutsplus.com/articles/how-to-use-the-contrast-checker-in-chrome-devtools--cms-31504
https://developer.chrome.com/docs/extensions/mv3/a11y/#colors
https://snook.ca/technical/colour_contrast/colour.html#fg=33FF33,bg=333333
http://www.color-blindness.com/coblis-color-blindness-simulator/

- [x] Sound
https://developer.chrome.com/docs/extensions/mv3/a11y/#sound
*No sounds used*

- [x] Images
*No images used in plugin*
https://developer.chrome.com/docs/extensions/mv3/a11y/#images
```
Provide informative alt text for images.
```
https://v1.grommet.io/docs/accessibility
```
Custom title
Grommet components can be read by screen readers. The default textual description varies for each component. In general, this description is generic and lacks contextual information that help users to understand the surrounding data. Most Grommet components support a property called a11yTitle that enables callers to provide a better description.

Example:
Using a11yTitle for icons to explain their purpose.

<Status a11yTitle='Server Down'
  value='critical' />
For this example, will read "Server Down, Image" instead of "Critical, Image".
```

## Grommet
https://v1.grommet.io/docs/accessibility
https://v2.grommet.io/

- [ ] Custom title
```
Using a11yTitle for icons to explain their purpose.

<Status a11yTitle='Server Down'
  value='critical' />
For this example, will read "Server Down, Image" instead of "Critical, Image".
```

- [ ] Language Support
```
Example
<App lang='en-US'>
  Testing
</App>
```
- [ ] Skip Links
```
Skip links will be automatically presented when pressing tab in a given page.
```

## React

- [ ] WAI-ARIA
https://reactjs.org/docs/accessibility.html#wai-aria
https://developer.chrome.com/docs/devtools/accessibility/reference/#computed
https://developer.chrome.com/docs/devtools/accessibility/reference/#pane
https://www.w3.org/WAI/intro/aria
https://www.w3.org/TR/wai-aria-practices/
https://www.w3.org/TR/wai-aria/#states_and_properties
https://www.w3.org/TR/wai-aria/#roles
https://www.w3.org/TR/wai-aria/#aria-label
https://www.w3.org/TR/wai-aria/#aria-labelledby

Use `aria-label` and `aria-labelledby` for non-text elements eg where `Name` attribute (as shown in developer tools Accessibility tab) is `""`

- [ ] Semantic HTML
https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML
https://reactjs.org/docs/accessibility.html#semantic-html

- [ ] Accessible Forms
https://reactjs.org/docs/accessibility.html#accessible-forms

- [x] Keyboard Support
https://reactjs.org/docs/accessibility.html#focus-control

## General

- [ ] Language Support
- https://webaim.org/techniques/language/
- https://www.i18next.com/overview/configuration-options#languages-namespaces-resources

## Final

- [ ] Verify Screen Reader use

- [x] Add precommit checks for accessibility issues
https://github.com/jsx-eslint/eslint-plugin-jsx-a11y
https://reactjs.org/docs/accessibility.html#development-assistance
https://elfi-y.medium.com/a-reference-manual-for-automating-accessibility-testing-in-react-gatsby-ii-31572927e027
https://blog.logrocket.com/build-robust-react-app-husky-pre-commit-hooks-github-actions/

[Edit 2] Restructured above content. Added Sections
[Edit 3] Added references for precommit
[Edit 4] Added reasoning for not testing Braille
